code/acting/src/acting/vehicle_controller.py

```diff
@@ -0,0 +1,217 @@
+                            "controller.")
+                raise Exception("Requested Controller not found")
+
+            message = CarlaEgoVehicleControl()
+            message.reverse = False
+            if self.__throttle > 0:  # todo: driving backwards?
+                message.brake = 0
+                message.throttle = self.__throttle
+            else:
+                message.throttle = 0
+                message.brake = abs(self.__throttle)
+
+            message.hand_brake = False
+            message.manual_gear_shift = False
+            message.steer = steer
+            # message.throttle = 5  # for testing todo: delete
+            message.gear = 1
+            message.header.stamp = roscomp.ros_timestamp(self.get_time(),
+                                                         from_sec=True)
+            self.control_publisher.publish(message)
+
+        self.new_timer(self.control_loop_rate, loop)
+        self.spin()
+
+    def __emergency_break(self, data) -> None:
+        """
+        Executes emergency stop
+        :param data:
+        :return:
+        """
+        if not data.data:  # not an emergency
+            return
+        if self.__emergency:  # emergency was already triggered
+            return
+        self.loginfo("Emergency breaking engaged")
+        self.__emergency = True
+        message = CarlaEgoVehicleControl()
+        message.throttle = 1
+        message.steer = 1  # todo: maybe use 30 degree angle
+        message.brake = 1
+        message.reverse = True
+        message.hand_brake = True
+        message.manual_gear_shift = False
+        message.header.stamp = roscomp.ros_timestamp(self.get_time(),
+                                                     from_sec=True)
+        self.control_publisher.publish(message)
+
+    def __get_velocity(self, data: CarlaSpeedometer) -> None:
+        """
+        Ends emergency if vehicle velocity reaches 0. Sends emergency msg
+        with data = False after stopping.
+        :param data:
+        :return:
+        """
+        if not self.__emergency:  # nothing to do in this case
+            return
+        if data.speed < 0.1:  # vehicle has come to a stop
+            self.__emergency = False
+            message = CarlaEgoVehicleControl()
+            message.throttle = 0
+            message.steer = 0
+            message.brake = 1
+            message.reverse = False
+            message.hand_brake = False
+            message.manual_gear_shift = False
+            message.header.stamp = roscomp.ros_timestamp(self.get_time(),
+                                                         from_sec=True)
+            self.control_publisher.publish(message)
+
+            self.loginfo("Emergency breaking disengaged "
+                         "(Emergency breaking has been executed successfully)")
+            self.emergency_pub.publish(
+                Bool(False))  # todo: publish multiple times?
```

Comment:
I think that's a very good point. In my opinion the emergency status should always be published with the highest sensible frequency.

Reply:
Das ist ja nur die Nachricht, dass der emergency vorbei ist. Das können wir auch nicht zu lange published, weil wir sonst eventuell Nachrichten überschreiben, die schon den nächsten emegency auslösen sollen. Außerdem sollte ROS glaube ich keine Nachrichten verschlucken

Reply:
Man könnte aber tatsächlich noch die `control_loop_rate` erhöhen. Keine Ahnung, ob das nötig/sinnvoll ist

Reply:
Man könnte aber tatsächlich noch die `control_loop_rate` erhöhen. Keine Ahnung, ob das nötig bzw. sinnvoll ist

Reply:
In 51f692b3a0e3fe1541e96c942793f537642e8f85 I changed the statement to send the msg multiple times. Does that resolve the issue for you?

Reply:
I also changed the the qos profile of the emergency msgs in 064d8d20897f6d424c8f1cf73ca6917f29a4f392